apps/src/codebridge/InfoPanel/ValidatedInstructions.tsx

```diff
@@ -0,0 +1,323 @@
+  beforeNextLevel?: () => void;
+  /** If the instructions panel should be rendered vertically or horizontally. Defaults to vertical. */
+  layout?: 'vertical' | 'horizontal';
+  /**
+   * A callback when the user clicks on clickable text.
+   */
+  handleInstructionsTextClick?: (id: string) => void;
+  /** Optional classname for the container */
+  className?: string;
+}
+
+interface NavigationButtonProps {
+  showNavigation: boolean;
+  navigationText: string;
+  handleNavigation: () => void;
+  navigationIcon?: FontAwesomeV6IconProps;
+}
+
+/**
+ * Instructions panel with built-in validation feedback.
+ * This instructions panel shows the validation results for a level, if they exist,
+ * and provides navigation once validation has passed.
+ */
+const ValidatedInstructions: React.FunctionComponent<InstructionsProps> = ({
+  beforeNextLevel,
+  layout = 'vertical',
+  handleInstructionsTextClick,
+  className,
+}) => {
+  const {onRun, onStop} = useCodebridgeContext();
+  const dialogControl = useDialogControl();
+
+  const instructionsText = useAppSelector(
+    state => state.lab.levelProperties?.longInstructions
+  );
+  const hasNextLevel = useSelector(state => nextLevelId(state) !== undefined);
+  const {hasConditions, satisfied} = useAppSelector(
+    state => state.lab.validationState
+  );
+  const predictSettings = useAppSelector(
+    state => state.lab.levelProperties?.predictSettings
+  );
+  const predictResponse = useAppSelector(state => state.predictLevel.response);
+  const predictAnswerLocked = useAppSelector(isPredictAnswerLocked);
+  const hasRun = useAppSelector(state => state.lab2System.hasRun);
+  const hasSubmitted = useAppSelector(
+    state => getCurrentLevel(state)?.status === LevelStatus.submitted
+  );
+  const isSubmittable = useAppSelector(
+    state => state.lab.levelProperties?.submittable
+  );
+  const source = useAppSelector(
+    state => state.lab2Project.projectSource?.source
+  ) as MultiFileSource | undefined;
+
+  const appType = useAppSelector(state => state.lab.levelProperties?.appName);
+  const isTesting = useAppSelector(state => state.lab2System.isTesting);
+  const isLoadingEnvironment = useAppSelector(
+    state => state.lab2System.loadingCodeEnvironment
+  );
+
+  const dispatch = useAppDispatch();
+
+  const {theme} = useContext(ThemeContext) || 'dark';
+
+  const vertical = layout === 'vertical';
+
+  const onFinish = () => {
+    // no op for now
+    // Tracked here: https://codedotorg.atlassian.net/browse/CT-664
+  };
+
+  const onNextPanel = useCallback(() => {
+    if (beforeNextLevel) {
+      beforeNextLevel();
+    }
+    dispatch(navigateToNextLevel());
+  }, [dispatch, beforeNextLevel]);
+
+  const onSubmit = () => {
+    const dialogTitle = hasSubmitted
+      ? commonI18n.unsubmitYourProject()
+      : commonI18n.submitYourProject();
+    const dialogMessage = hasSubmitted
+      ? commonI18n.unsubmitYourProjectConfirm()
+      : commonI18n.submitYourProjectConfirm();
```

Comment:
This is fine as is, but I always hate repeated conditionals. I might set it up something like this:

```
const dialogStrings = new Map(
  [
    [ hasSubmitted, {
      title : commonI18n.unsubmitYourProject(),
      message : commonI18n. unsubmitYourProjectConfirm()
    }],
    [ !hasSubmitted, {
      title : commonI18n.submitYourProject(),
      message : commonI18n.submitYourProjectConfirm()
    }]
  ]
);

const [dialogTitle, dialogMessage] = dialogStrings.get(hasSubmitted);
```

the `dialogStrings` constant could even be moved elsewhere outside the component, if beneficial.

It's using a wordier `Map` instead of just an object so we don't end up casting our boolean keys into strings.

Is this beneficial enough to bother swapping to the lookup just to avoid the duplicated conditional? I dunno. Dealer's choice.

Reply:
I personally find the map option harder to read, so I'll leave as-is for now.